tfhe/src/shortint/engine/server_side/mod.rs

```diff
@@ -482,10 +482,22 @@ impl ShortintEngine {
     where
         F: Fn(u64, u64) -> u64,
     {
-        let mut ct_res = ct_left.clone();
+        // Generate the lookup table for the function
+        let factor = MessageModulus(ct_right.degree.0 + 1);
+        let lookup_table =
+            self.generate_lookup_table_bivariate_with_factor(server_key, f, factor)?;
 
-        self.smart_evaluate_bivariate_function_assign(server_key, &mut ct_res, ct_right, f)?;
-        Ok(ct_res)
+        if !server_key.is_functional_bivariate_pbs_possible(ct_left, ct_right) {
+            // After the message_extract, we'll have ct_left, ct_right in [0, message_modulus[
+            // so the factor has to be message_modulus
+            assert_eq!(ct_right.message_modulus.0, lookup_table.ct_right_modulus.0);
+            self.message_extract_assign(server_key, ct_left)?;
+            self.message_extract_assign(server_key, ct_right)?;
+        }
```

Comment:
I think it would be easier to generate the Lut *after* the potential message extracts have been done, so that the assert is not needed.

Ie why chose a factor if it may be wrong, when you could choose the factor and have it be always valid ?

Reply:
that I did it in a rush and to not take this for the best solution, if there is a better piece of code use it, I would not know if mine makes much sense

Reply:
Fixed